packages/block-editor/src/store/reducer.js

```diff
@@ -1981,11 +1981,20 @@ export function blockEditingModes( state = new Map(), action ) {
 	switch ( action.type ) {
 		case 'SET_BLOCK_EDITING_MODE':
 			return new Map( state ).set( action.clientId, action.mode );
+		case 'SET_BLOCK_EDITING_MODES':
+			return new Map( ...state, ...action.modes );
```

Comment:
> May it be the individual toggling of modes @ellatrix mentioned?

Yeah, it looked like we were constructing a new Map every time we updated one of them which was causing some memory thrashing. I added this batch action to update a bunch at once which fixed that.